File: app/src/main/java/com/timotiusoktorio/pencake/ui/adapters/CartRvAdapter.java
package com.timotiusoktorio.pencake.ui.adapters;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.squareup.picasso.Picasso;
import com.timotiusoktorio.pencake.R;
import com.timotiusoktorio.pencake.data.model.CartItem;
import com.timotiusoktorio.pencake.data.model.Product;
import com.timotiusoktorio.pencake.databinding.ListItemCartBinding;

import java.util.List;

public class CartRvAdapter extends RecyclerView.Adapter<CartRvAdapter.ViewHolder> {

    private List<CartItem> cart;
    private OnItemClickListener listener;

    public CartRvAdapter(List<CartItem> cart) {
        this.cart = cart;
    }

    public void setListener(OnItemClickListener listener) {
        this.listener = listener;
    }

    public void updateData(List<CartItem> data) {
        cart.clear();
        cart.addAll(data);
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        ListItemCartBinding binding = ListItemCartBinding.inflate(inflater, parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(cart.get(position));
    }

    @Override
    public int getItemCount() {
        return cart.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        private ListItemCartBinding binding;

        ViewHolder(ListItemCartBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(final CartItem cartItem) {
            Product product = Product.fromJson(cartItem.getProductJson());
            binding.setCartItem(cartItem);
            binding.setProduct(product);

            Picasso.get().load(product.getImageUrls().get(0)).fit().centerCrop()
                    .placeholder(R.drawable.placeholder_image)
                    .into(binding.productPhotoIv);

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (listener != null) {
                        listener.onItemClick(cartItem);
                    }
                }
            });
        }
    }

    public interface OnItemClickListener {

        void onItemClick(CartItem cartItem);
    }
}

File: app/src/main/java/com/timotiusoktorio/pencake/ui/adapters/CartRvAdapter.kt
package com.timotiusoktorio.pencake.ui.adapters

import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import com.timotiusoktorio.pencake.R
import com.timotiusoktorio.pencake.data.model.CartItem
import com.timotiusoktorio.pencake.data.model.Product
import com.timotiusoktorio.pencake.extensions.inflate
import com.timotiusoktorio.pencake.extensions.loadUrl
import kotlinx.android.synthetic.main.list_item_cart.view.*

class CartRvAdapter(private val cart: MutableList<CartItem> = mutableListOf(),
                    private val listener: (CartItem) -> Unit) : RecyclerView.Adapter<CartRvAdapter.ViewHolder>() {

    fun updateData(data: List<CartItem>) {
        cart.clear()
        cart.addAll(data)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder = ViewHolder(parent.inflate(R.layout.list_item_cart))

    override fun onBindViewHolder(holder: ViewHolder, position: Int) = holder.bind(cart[position])

    override fun getItemCount(): Int = cart.size

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(cartItem: CartItem) = with(itemView) {
            val product = Product.fromJson(cartItem.productJson)
            productNameTv.text = product.name
            quantityTv.text = context.getString(R.string.string_format_quantity, cartItem.quantity)
            productSizeTv.text = product.prices[cartItem.selectedSizeIndex].size
            specialRequestsTv.text = if (cartItem.specialRequests.isNullOrEmpty()) context.getString(R.string.text_no_special_requests) else cartItem.specialRequests
            subtotalTv.text = context.getString(R.string.string_format_price, cartItem.subtotal)
            productPhotoIv.loadUrl(product.imageUrls[0])
            setOnClickListener { listener(cartItem) }
        }
    }
}